### PR TITLE
mgr/dashboard: Display cluster status in favicon

### PR DESCRIPTION
**Pull request by tspmelo**
Fixes: https://tracker.ceph.com/issues/44542

Signed-off-by: Tiago Melo <tmelo@suse.com>

image

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tspmelo** [comment]
jenkins test dashboard

**Comment by tspmelo** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      switch (status) {
+        case 'HEALTH_OK':
+          color = '#00bb00';
+          break;
+        case 'HEALTH_WARN':
+          color = '#ffa500';
+          break;
+        case 'HEALTH_ERR':
+          color = '#ff0000';
+          break;
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts:63]: Can colors be extracted to a constants file, just no to have the same colors repeated (e.g.: HealthPipe, ...)?

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts:63]: sure, I can do that.

**Comment by tspmelo** [comment]
@epuertat I also found some package that did it, not sure if it was that one. 
But since this was a simple functionality I didn't want to add another external dependency where we would only use part of it.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts (hunk 1 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts]: Thanks!

**Comment by tspmelo** [comment]
jenkins test dashboard

**Comment by tspmelo** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    const context = canvas.getContext('2d');
+    const img = this.document.createElement('img');
+    img.src = this.url;
+
+    img.onload = () => {
+      // Draw Original Favicon as Background
+      context.drawImage(img, 0, 0, faviconSize, faviconSize);
+
+      // Draw Notification Circle
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts:50]: I would like to suggest to cut-out the circle area with radius + 1 for a better look.

Before:
screenshot

After:
screenshot2

screenshot3

```suggestion
      // Cut notification circle area
      context.save();
      context.globalCompositeOperation = 'destination-out';
      context.beginPath();
      context.arc(canvas.width - faviconSize / 4, faviconSize / 4, (faviconSize / 4) + 1, 0, 2 * Math.PI);
      context.fill();
      context.restore();

      // Draw notification circle
```

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/favicon.service.ts]: Done. Went with 1px.

**Comment by tspmelo** [comment]
jenkins test dashboard backend

**Comment by LenzGr** [comment]
Did someone test this on Firefox? It does not work for me: https://tracker.ceph.com/issues/46918

**Pull request merged**